### DISCUSSION TITLE
Is "Minimum" Test supported by Instancio?

### DISCUSSION THREAD

**Discussion by gizmo1177** [instancio/instancio#809] [Q&A] [answered]
We evaluated Instancio and it works fine. What we call a "Maximum" - Test is supported and validation constraints are respected. Means ALL the properties will be initialized.

Is the opposite also supported somehow? (Minimum Test). Means the data generator only prefills the required fields and all the other properties stays null (empty). Thanks a lot.

**Comment by armandino** [comment]
The way Instancio works is it attempts to fully-populate an object, letting the user specify which fields to `ignore()` or `set()` to null.

For your second use-case, how would the library differentiate between required and optional fields? Can you provide a POJO as an example?

**Reply by gizmo1177** [reply]
Thanks for your answer. Fully populating an object and respecting the validation annotations (if wanted) is very nice and works as expected. By populating only the minimum needed data we want to find Null Pointer Exceptions and so on. It would mean that references/fields that are not required (e.g. have no `@NotNull` annotation) will simply not initialized and stay null.

**Example:**

```java
public class A {
  private B b;
  @ Size(max = 200)
  private String nullable;
  @ NotNull
  private Sting notNullable;
}

public class B {
  ...
}
```

- If we generate the data for class `A` in the "Maximum" case all the fields are popuplated and in case Bean Validation is enabled the rules are respected. This is already the case.
- If we generate the data for class `A` in the "Minimum" case only the fields/references that have a `@NotNull` constraint will be initialized by the generator. So the bean would be perfectly valid (according to the constraints) and we could use that instance in our tests to find possible NPE´s. I guess this is currently not supported but would be "nice to have". In the example above it could be to detect wrong usage of the property "nullable" or reference `B` in class `A`.

**Reply by armandino** [reply, marked as answer]
I see, thanks for the example. It should be possible to set all fields that do not have a `NotNull` annotation to null:

```java
A result = Instancio.of(A.class)
    .set(fields(field -> field.getDeclaredAnnotation(NotNull.class) == null), null)
    .create();
```

This uses a predicate selector `Select.fields(Predicate<Field>)` 

- https://javadoc.io/doc/org.instancio/instancio-core/latest/org/instancio/Select.html#fields(java.util.function.Predicate)

You could also use `ignore(fields(field -> field.getDeclaredAnnotation(NotNull.class) == null))` instead of `set()`. However, using `set()` would allow you to override certain fields to be not null if needed, e.g.

```java
A result = Instancio.of(A.class)
    .set(fields(field -> field.getDeclaredAnnotation(NotNull.class) == null), null)
    .set(field(A::getNullable), "some non-null value") // you can still set non-null value
    .create();
```

whereas if you use `ignore()` it wouldn't be possible (ignored fields can't set).

**Reply by gizmo1177** [reply]
Thanks a lot. This seems to be exactly what we need.
